Update the README (#925)
huggingface/optimum: README.md

| [Graphcore IPU](https://www.graphcore.ai/products/ipu)                                                                 | `python -m pip install optimum[graphcore]`        |
| [Habana Gaudi Processor (HPU)](https://habana.ai/training/)                                                            | `python -m pip install optimum[habana]`           |

To install from source:

```bash
python -m pip install git+https://github.com/huggingface/optimum.git
```

For the accelerator-specific features, append `#egg=optimum[accelerator_type]` to the above command:

```bash
python -m pip install git+https://github.com/huggingface/optimum.git#egg=optimum[onnxruntime]

> ```

Author: I would rather have all the Installation section together. I think it is small enough to keep it here. 
Or could we move the installation section to the bottom?

> ```

Reviewer: It's probably OK as is then!

## Accelerated Inference

🤗 Optimum provides multiple tools to export and run optimized models on various ecosystems: 

- ONNX / ONNX Runtime
- TensorFlow Lite

> - OpenVINO

Reviewer: You could add links to the doc with `[]()` here

While everything can be done in a programmatic way for a more cutomized experience, we provide a command-line interface to perform common tasks easily. 

### ONNX + ONNX Runtime

> It is possible to export 🤗 Transformers models to the [ONNX](https://onnx.ai/) format and perform graph optimization as well as quantization easily:

Author: Yes it works, about speedups I do not know. 
Let's wait for Ilyas to join.

Also will update that when I merge #927

```plain
optimum-cli export onnx -m deepset/roberta-base-squad2 --optimize O2 roberta_base_qa
```

The model can then be quantized using `onnxruntime`:

```bash
optimum-cli onnxruntime quantize --avx512 --onnx_model roberta_base_qa
```

These commands will export `deepset/roberta-base-squad2` perform [O2 graph optimization](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/optimization#optimization-configuration) on the exported model, and finally quantize it.

For more information on the ONNX export, please check the [documentation](https://huggingface.co/docs/optimum/exporters/onnx/usage_guides/export_a_model).

> #### Run the exported model using ONNX Runtime

Reviewer: I can work on adding a table/plot highlighting why one would want to run inference with ORT

> Once the model exported to the ONNX format, we provide Python classes enabling you to run the exported file in a seemless manner using [ONNX Runtime](https://onnxruntime.ai/) in the backend:

Reviewer: ```suggestion
Once the model is exported to the ONNX format, we provide Python classes enabling you to run the exported file in a seemless manner using [ONNX Runtime](https://onnxruntime.ai/) in the backend:
```

```python
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForQuestionAnswering

> model_name = "roberta_base_qa"

Reviewer: Either `roberta_base_qa_onnx/` or `roberta-base` with `export=True` IMO

Author: Nope, I basically take the same folder as the export command above.

```
# Run with ONNX Runtime.
outputs = ort_model(**inputs)

In this example, we've quantized a model from the Hugging Face Hub, in the same manner we can quantize a model hosted locally by providing the path to the directory containing the model weights. The result from applying the `quantize()` method is a `model_quantized.onnx` file that can be used to run inference. Here's an example of how to load an ONNX Runtime model and generate predictions with it:
answer_start_index = outputs.start_logits.argmax()
answer_end_index = outputs.end_logits.argmax()

```python
from optimum.onnxruntime import ORTModelForSequenceClassification
from transformers import pipeline, AutoTokenizer

model = ORTModelForSequenceClassification.from_pretrained(save_directory, file_name="model_quantized.onnx")
tokenizer = AutoTokenizer.from_pretrained(save_directory)
classifier = pipeline("text-classification", model=model, tokenizer=tokenizer)
results = classifier("I love burritos!")
predict_answer_tokens = inputs.input_ids[0, answer_start_index : answer_end_index + 1]
answer = tokenizer.decode(predict_answer_tokens, skip_special_tokens=True)
```

> You can find more examples in the [documentation](https://huggingface.co/docs/optimum/onnxruntime/quickstart) and in the [examples](https://github.com/huggingface/optimum/tree/main/examples/onnxruntime).

Reviewer 2: Maybe we can just link to the guide in the doc here: https://huggingface.co/docs/optimum/main/en/onnxruntime/usage_guides/models
Now I realize we don't have a simple example in the doc for inference with ORT with an already exported model.

> Just as for ONNX, it is possible to export to [TensorFlow Lite](https://www.tensorflow.org/lite) and quantize them:

Reviewer 2: ```suggestion
Just as for ONNX, it is possible to export models to [TensorFlow Lite](https://www.tensorflow.org/lite) and quantize them:
```

> optimum-cli export tflite -m deepset/roberta-base-squad2 --sequence_length 384  --quantize int8-dynamic roberta_tflite_model

Reviewer: Can you break line so that the full command is visible without right scrolling?

Author: I tried but then the command does not run by simply copy-pasting, do you have a suggestion?

Reviewer: You can do something like:
```
optimum-cli export onnx \
    --model gpt2 \
    gpt2_onnx/
```

Or maybe since it's subprocess under the hood this is failing for tflite?

Author: It worked thanks! I might have put spaces after the `\`...

> ### OpenVINO

Reviewer: Maybe note that `optimum-intel` must be installed

> To load a model and run inference with [OpenVINO Runtime](https://docs.openvino.ai/latest/home.html), you can just replace your `AutoModelForXxx` class with the corresponding `OVModelForXxx` class.

Reviewer: It would be cool to highlight why one would want to use OpenVINO @echarlaix (e.g. a table with speedups)

Reviewer: Maybe it can be in a following PR indeed

> If you want to load a PyTorch checkpoint, set `from_transformers=True` to convert your model to the OpenVINO IR (Intermediate Representation).

Reviewer: ```suggestion
If you want to load a PyTorch checkpoint, set `export=True` to convert your model to the OpenVINO IR (Intermediate Representation).
```

@echarlaix is this correct?

Author: @echarlaix ?

Reviewer: @michaelbenayoun I think it is

Author: Yes, from this [page](https://huggingface.co/docs/optimum/intel/inference) it seems like it. Updating that and merging.

platforms.
### ONNX + ONNX Runtime

Optimum enables the usage of popular compression techniques such as quantization and pruning by supporting [ONNX Runtime](https://onnxruntime.ai/docs/) along with Intel [Neural Compressor](https://www.intel.com/content/www/us/en/developer/tools/oneapi/neural-compressor.html) and OpenVINO [NNCF](https://docs.openvino.ai/latest/tmo_introduction.html).
It is possible to export 🤗 Transformers models to the ONNX format and perform graph optimization as well as quantization easily:

| Features                           | ONNX Runtime          |     Neural Compressor   |         OpenVINO        |
|:----------------------------------:|:---------------------:|:-----------------------:|:-----------------------:|
| Post-training Dynamic Quantization |  :heavy_check_mark:   |    :heavy_check_mark:   |          N/A            |
| Post-training Static Quantization  |  :heavy_check_mark:   |    :heavy_check_mark:   |    :heavy_check_mark:   |
| Quantization Aware Training (QAT)  |  Stay tuned! :star:   |    :heavy_check_mark:   |    :heavy_check_mark:   |

> | Pruning                            |        N/A            |    :heavy_check_mark:   |    :heavy_check_mark:   |

Author: I do not think this is needed here, it should be more in the documentation. 
Or otherwise I think we should be showing instead of telling via a small command example.

> | Pruning                            |        N/A            |    :heavy_check_mark:   |    :heavy_check_mark:   |

Reviewer 3: We cannot add an example for each, it would make the README overcrowded and unclear. The README is the perfect place to present the different features of each subpackages, in order for users to be redirected towards what they need. We also should have it in the documentation.

> | Pruning                            |        N/A            |    :heavy_check_mark:   |    :heavy_check_mark:   |

Author: Done

You can find more examples in the [documentation](https://huggingface.co/docs/optimum/intel/inference) and in the [examples](https://github.com/huggingface/optimum-intel/tree/main/examples/openvino).

> 

Author: Need to check internally.

You can find more examples in the [documentation](https://huggingface.co/docs/optimum/intel/inference) and in the [examples](https://github.com/huggingface/optimum-intel/tree/main/examples/openvino).

### Features summary

<!--

| Features                           | ONNX Runtime          |     Neural Compressor   |         OpenVINO        |         TensorFlow Lite |
|:----------------------------------:|:---------------------:|:-----------------------:|:-----------------------:|:-----------------------:|
| Graph optimization                 |  :heavy_check_mark:   |    N/A                  |          N/A            |     N/A                 |
| Post-training Dynamic Quantization |  :heavy_check_mark:   |    :heavy_check_mark:   |          N/A            |     :heavy_check_mark:  |
| Post-training Static Quantization  |  :heavy_check_mark:   |    :heavy_check_mark:   |    :heavy_check_mark:   |    :heavy_check_mark:   |

> | Quantization Aware Training (QAT)  |  Stay tuned! :star:   |    :heavy_check_mark:   |    :heavy_check_mark:   |    :heavy_check_mark:   |

Reviewer: ```suggestion
| Quantization Aware Training (QAT)  |  Unsupported   |    :heavy_check_mark:   |    :heavy_check_mark:   |    :heavy_check_mark:   |
```

| Floating-point 16                  |  :heavy_check_mark:   | N/A                     |    N/A                  |     :heavy_check_mark:  |
| Pruning                            |        N/A            |    :heavy_check_mark:   |    :heavy_check_mark:   |    N/A                  |

-->

| Features                           | ONNX Runtime       | Neural Compressor  | OpenVINO           | TensorFlow Lite    |
|:----------------------------------:|:------------------:|:------------------:|:------------------:|:------------------:|
| Graph optimization                 | :heavy_check_mark: | N/A                | N/A                | N/A                |
| Post-training Dynamic Quantization | :heavy_check_mark: | :heavy_check_mark: | N/A                | :heavy_check_mark: |
| Post-training Static Quantization  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

> | Quantization Aware Training (QAT)  | Stay tuned! :star: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

Reviewer 3: ```suggestion
| Quantization Aware Training (QAT)  | N/A | :heavy_check_mark: | :heavy_check_mark: | N/A |
```

> | Floating-point 16                  | :heavy_check_mark: | N/A                | N/A                | :heavy_check_mark: |

Reviewer: ```suggestion
| fp16                  | :heavy_check_mark: | N/A                | N/A                | :heavy_check_mark: |
```

| Pruning                            | N/A                | :heavy_check_mark: | :heavy_check_mark: | N/A                |

## Accelerated training

🤗 Optimum provides wrappers around the original 🤗 Transformers [Trainer](https://huggingface.co/docs/transformers/main_classes/trainer) to enable training on powerful hardware easily.
We support many providers:

- Habana's Gaudi processors
- Graphcore's IPUs 

> - ONNX Runtime TODO <= not a provider per-se, it is more software?

Reviewer 2: ```suggestion
- ONNX Runtime (optimized for GPU)
```
Maybe we can just say that it is for GPUs?

> - ONNX Runtime TODO <= not a provider per-se, it is more software?

Author: Can be done in another PR, the plan for this one was to make it less fluffy and more to the point.

> - ONNX Runtime TODO <= not a provider per-se, it is more software?

Reviewer: Sounds good!
